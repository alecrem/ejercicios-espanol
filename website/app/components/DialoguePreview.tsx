interface DialoguePreviewProps {
  content: string;
}

const styles = {
  dialogue: {
    lineHeight: "1.8",
    fontSize: "1.1rem",
  },
  speaker: {
    fontWeight: "600",
    color: "#1a1a1a",
  },
  line: {
    marginBottom: "0.8rem",
  },
};

function parseDialogue(content: string) {
  const lines = content.split('\n').filter(line => line.trim());
  const dialogueLines: Array<{ speaker: string; text: string }> = [];
  
  let title = "";
  let subtitle = "";
  
  for (const line of lines) {
    // Skip markdown headers
    if (line.startsWith('#')) {
      if (line.startsWith('# ')) {
        title = line.replace('# ', '');
      } else if (line.startsWith('## ')) {
        subtitle = line.replace('## ', '');
      }
      continue;
    }
    
    // Parse dialogue lines with **Speaker:** format
    const speakerMatch = line.match(/\*\*([^*]+):\*\*\s*(.*)/);
    if (speakerMatch) {
      const [, speaker, text] = speakerMatch;
      dialogueLines.push({ speaker, text });
    }
  }
  
  return { title, subtitle, dialogueLines };
}

export default function DialoguePreview({ content }: DialoguePreviewProps) {
  const { title, subtitle, dialogueLines } = parseDialogue(content);
  
  return (
    <div>
      <h2 style={{ 
        fontSize: "1.5rem", 
        fontWeight: "700", 
        marginBottom: "1rem",
        fontFamily: "'Delius', system-ui, sans-serif",
      }}>
        {title}
      </h2>
      
      {subtitle && (
        <h3 style={{
          fontSize: "1.2rem",
          fontWeight: "600", 
          marginBottom: "1.5rem",
          color: "#4b5563",
        }}>
          {subtitle}
        </h3>
      )}
      
      <div style={styles.dialogue}>
        {dialogueLines.map((line, index) => (
          <div key={index} style={styles.line}>
            <span style={styles.speaker}>{line.speaker}:</span>{' '}
            <span>{line.text}</span>
          </div>
        ))}
      </div>
    </div>
  );
}