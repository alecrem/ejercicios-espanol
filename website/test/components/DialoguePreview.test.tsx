import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import DialoguePreview from "~/components/DialoguePreview";

describe("DialoguePreview", () => {
  const mockContent = `# Un día en casa

## Conversación entre Ana y Carlos

**Ana:** ¡Hola Carlos! ¿Cómo **estás** hoy?

**Carlos:** ¡Hola Ana! Estoy **muy bien**, gracias. ¿Y tú?

**Ana:** Yo también estoy bien. ¿Qué planes **tienes** para hoy?

**Carlos:** Voy a **estudiar** español. ¿Te gustaría estudiar conmigo?`;

  it("displays title from markdown header", () => {
    render(<DialoguePreview content={mockContent} />);
    expect(screen.getByRole("heading", { level: 2, name: "Un día en casa" })).toBeInTheDocument();
  });

  it("displays subtitle from markdown subheader", () => {
    render(<DialoguePreview content={mockContent} />);
    expect(screen.getByRole("heading", { level: 3, name: "Conversación entre Ana y Carlos" })).toBeInTheDocument();
  });

  it("parses and displays dialogue lines with speakers", () => {
    render(<DialoguePreview content={mockContent} />);
    
    expect(screen.getAllByText("Ana:")).toHaveLength(2);
    expect(screen.getAllByText("Carlos:")).toHaveLength(2);
    expect(screen.getByText(/¡Hola Carlos! ¿Cómo/)).toBeInTheDocument();
    expect(screen.getByText(/Voy a/)).toBeInTheDocument();
  });

  it("renders bold text correctly", () => {
    render(<DialoguePreview content={mockContent} />);
    
    expect(screen.getByText("estás")).toBeInTheDocument();
    expect(screen.getByText("muy bien")).toBeInTheDocument();
    expect(screen.getByText("tienes")).toBeInTheDocument();
    expect(screen.getByText("estudiar")).toBeInTheDocument();
  });

  it("handles content without subtitle", () => {
    const contentNoSubtitle = `# Solo título

**Ana:** Hola mundo.

**Carlos:** ¡Hola!`;

    render(<DialoguePreview content={contentNoSubtitle} />);
    
    expect(screen.getByRole("heading", { level: 2, name: "Solo título" })).toBeInTheDocument();
    expect(screen.queryByRole("heading", { level: 3 })).not.toBeInTheDocument();
  });

  it("handles empty content gracefully", () => {
    render(<DialoguePreview content="" />);
    
    // Should render without crashing - component will render an empty title
    const container = document.body;
    expect(container).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<DialoguePreview content={mockContent} />);
    expect(container).toMatchSnapshot();
  });

  it("filters out empty lines", () => {
    const contentWithEmptyLines = `# Test Title


**Ana:** First line.



**Carlos:** Second line.


`;

    render(<DialoguePreview content={contentWithEmptyLines} />);
    
    // Should only show the two dialogue lines
    expect(screen.getAllByText("Ana:")).toHaveLength(1);
    expect(screen.getAllByText("Carlos:")).toHaveLength(1);
    expect(screen.getByText("First line.")).toBeInTheDocument();
    expect(screen.getByText("Second line.")).toBeInTheDocument();
  });
});