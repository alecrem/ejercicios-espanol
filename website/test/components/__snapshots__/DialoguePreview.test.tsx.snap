// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DialoguePreview > matches snapshot 1`] = `
<div>
  <div>
    <h2
      style="font-size: 1.5rem; font-weight: 700; margin-bottom: 1rem; font-family: "Delius", system-ui, sans-serif;"
    >
      Un día en casa
    </h2>
    <h3
      style="font-size: 1.2rem; font-weight: 600; margin-bottom: 1.5rem; color: rgb(75, 85, 99);"
    >
      Conversación entre Ana y Carlos
    </h3>
    <div
      style="line-height: 1.8; font-size: 1.1rem;"
    >
      <div
        style="margin-bottom: 0.8rem;"
      >
        <span
          style="font-weight: 600; color: rgb(26, 26, 26);"
        >
          Ana
          :
        </span>
         
        <span>
          ¡Hola Carlos! ¿Cómo 
          <strong>
            estás
          </strong>
           hoy?
        </span>
      </div>
      <div
        style="margin-bottom: 0.8rem;"
      >
        <span
          style="font-weight: 600; color: rgb(26, 26, 26);"
        >
          Carlos
          :
        </span>
         
        <span>
          ¡Hola Ana! Estoy 
          <strong>
            muy bien
          </strong>
          , gracias. ¿Y tú?
        </span>
      </div>
      <div
        style="margin-bottom: 0.8rem;"
      >
        <span
          style="font-weight: 600; color: rgb(26, 26, 26);"
        >
          Ana
          :
        </span>
         
        <span>
          Yo también estoy bien. ¿Qué planes 
          <strong>
            tienes
          </strong>
           para hoy?
        </span>
      </div>
      <div
        style="margin-bottom: 0.8rem;"
      >
        <span
          style="font-weight: 600; color: rgb(26, 26, 26);"
        >
          Carlos
          :
        </span>
         
        <span>
          Voy a 
          <strong>
            estudiar
          </strong>
           español. ¿Te gustaría estudiar conmigo?
        </span>
      </div>
    </div>
  </div>
</div>
`;
